File: src/discovery.rs
//! Finding the Git repositories below a directory.

use std::collections::{HashSet, VecDeque};
use std::ffi::OsString;
use std::fs;
use std::io;
use std::path::{Path, PathBuf};

/// One repository that was found.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Repository {
    /// The name of the directory, as it appears on disk.
    pub name: OsString,
    /// The absolute path of the directory.
    pub path: PathBuf,
}

/// The outcome of a search.
#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub struct Discovery {
    /// The repositories that were found, sorted by name.
    pub repositories: Vec<Repository>,
    /// Directories that could not be read, in the order they were visited.
    pub warnings: Vec<String>,
}

/// Whether `path` itself is a repository, that is whether it holds a `.git`
/// directory or a `.git` file (worktrees and submodules use the latter).
pub fn is_repository(path: &Path) -> bool {
    match fs::metadata(path.join(".git")) {
        Ok(metadata) => metadata.is_dir() || metadata.is_file(),
        Err(_) => false,
    }
}

/// Search `root` for repositories, up to `max_depth` levels below it.
///
/// Level 1 is a direct child of `root`. Symlinked directories are followed and
/// the traversal never descends into a repository, so nested checkouts such as
/// submodules are not reported twice. A repository that is reachable through
/// several names (a symbolic link next to its target, for example) is reported
/// for every name, exactly like the `*/` glob of the shell version did.
pub fn discover(root: &Path, max_depth: usize) -> io::Result<Discovery> {
    let mut discovery = Discovery::default();
    if max_depth == 0 {
        return Ok(discovery);
    }

    let mut visited: HashSet<PathBuf> = HashSet::new();
    if let Ok(canonical) = fs::canonicalize(root) {
        visited.insert(canonical);
    }

    let mut queue: VecDeque<(PathBuf, usize)> = VecDeque::new();
    queue.push_back((root.to_path_buf(), 1));

    while let Some((directory, level)) = queue.pop_front() {
        let entries = match read_entries(&directory) {
            Ok(entries) => entries,
            Err(error) if directory == root => return Err(error),
            Err(error) => {
                discovery.warnings.push(format!(
                    "mgit: cannot read `{}`: {error}\n",
                    directory.display()
                ));
                continue;
            }
        };

        for entry in entries {
            let path = entry;
            if !is_directory(&path) {
                continue;
            }

            if is_repository(&path) {
                discovery.repositories.push(Repository {
                    name: file_name(&path),
                    path,
                });
                continue;
            }

            // Walking into the same real directory twice would loop forever on
            // symbolic links, so every directory is only visited once.
            let identity = fs::canonicalize(&path).unwrap_or_else(|_| path.clone());
            if level < max_depth && visited.insert(identity) {
                queue.push_back((path, level + 1));
            }
        }
    }

    discovery
        .repositories
        .sort_by_key(|repository| sort_key(&repository.name));
    Ok(discovery)
}

/// Read a directory and return its entries sorted by file name.
fn read_entries(directory: &Path) -> io::Result<Vec<PathBuf>> {
    let mut entries: Vec<PathBuf> = fs::read_dir(directory)?
        .map(|entry| entry.map(|entry| entry.path()))
        .collect::<io::Result<_>>()?;
    entries.sort_by(|left, right| left.file_name().cmp(&right.file_name()));
    Ok(entries)
}

fn is_directory(path: &Path) -> bool {
    fs::metadata(path)
        .map(|metadata| metadata.is_dir())
        .unwrap_or(false)
}

fn file_name(path: &Path) -> OsString {
    path.file_name()
        .map(OsString::from)
        .unwrap_or_else(|| OsString::from(path.as_os_str()))
}

/// Names are compared without their case so that the order looks natural on
/// case insensitive file systems, and the exact spelling breaks ties.
fn sort_key(name: &OsString) -> (String, String) {
    let lossy = name.to_string_lossy().into_owned();
    let folded: String = lossy
        .chars()
        .flat_map(|character| character.to_lowercase())
        .collect();
    (folded, lossy)
}

File: website/assets/js/engine/discovery.js
// Finding the Git repositories below a directory, ported from src/discovery.rs.
//
// Breadth first, one level at a time; a directory that is a repository is
// recorded and never descended into; symbolic links are followed but every
// real directory is only entered once, so a link loop cannot spin forever.
// The result is sorted by name, case insensitively, with the exact spelling
// breaking ties, which is what makes the order identical on macOS, Linux and
// Windows.

import { baseName } from './fs.js';

/**
 * @param {import('./fs.js').Filesystem} fs
 * @param {string} root Absolute path of the directory to search.
 * @param {number} maxDepth
 * @returns {{repositories: {name: string, path: string}[], warnings: string[], trace: TraceStep[], error?: string}}
 */
export function discover(fs, root, maxDepth) {
  /** @type {{name: string, path: string}[]} */
  const repositories = [];
  /** @type {string[]} */
  const warnings = [];
  /** @type {TraceStep[]} */
  const trace = [];

  if (maxDepth === 0) return { repositories, warnings, trace };

  const visited = new Set([fs.canonical(root)]);
  /** @type {{path: string, level: number}[]} */
  const queue = [{ path: root, level: 1 }];

  while (queue.length > 0) {
    const { path, level } = queue.shift();
    const listing = fs.readDir(path);
    if (!listing.ok) {
      if (path === root) return { repositories: [], warnings, trace, error: listing.error };
      warnings.push('mgit: cannot read `' + path + '`: ' + listing.error + '\n');
      trace.push({ kind: 'unreadable', path, level, detail: listing.error });
      continue;
    }

    trace.push({ kind: 'scan', path, level, detail: String(listing.entries.length) });

    for (const entry of listing.entries) {
      if (!fs.isDirectory(entry.path)) {
        trace.push({ kind: 'skip-file', path: entry.path, level });
        continue;
      }

      if (fs.isRepository(entry.path)) {
        repositories.push({ name: entry.name, path: entry.path });
        trace.push({ kind: 'repository', path: entry.path, level });
        continue;
      }

      const identity = fs.canonical(entry.path);
      if (level >= maxDepth) {
        trace.push({ kind: 'depth-limit', path: entry.path, level });
        continue;
      }
      if (visited.has(identity)) {
        trace.push({ kind: 'already-visited', path: entry.path, level, detail: identity });
        continue;
      }
      visited.add(identity);
      queue.push({ path: entry.path, level: level + 1 });
      trace.push({ kind: 'descend', path: entry.path, level });
    }
  }

  repositories.sort((left, right) => compareKeys(sortKey(left.name), sortKey(right.name)));
  return { repositories, warnings, trace };
}

/** Whether the directory itself is a repository. */
export function isRepository(fs, path) {
  return fs.isRepository(path);
}

function sortKey(name) {
  return [name.toLowerCase(), name];
}

function compareKeys(left, right) {
  if (left[0] !== right[0]) return left[0] < right[0] ? -1 : 1;
  if (left[1] !== right[1]) return left[1] < right[1] ? -1 : 1;
  return 0;
}

/**
 * @typedef {object} TraceStep
 * @property {'scan'|'repository'|'descend'|'depth-limit'|'already-visited'|'skip-file'|'unreadable'} kind
 * @property {string} path
 * @property {number} level
 * @property {string} [detail]
 */

export { baseName };
